### PR TITLE
Disallow void branches

### LINKED ISSUE #7256: Checking for void in logical operations
In many cases it is not desirable to allow comparison with undefined. Is it somehow enforceable that this will be detected during compilation?

``` javascript
function a() { }

if (a()) { } // this should throw error
```

I guess this is by design not available.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #7256
#### Open questions:

Should we cover `null`/`undefined` types?
#### TODO:
- Conditional expressions
- Negation expressions.

### CODE REVIEW

File: tests/baselines/reference/logicalAndOperatorWithEveryType.errors.txt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+tests/cases/conformance/expressions/binaryOperators/logicalAndOperator/logicalAndOperatorWithEveryType.ts(85,11): error TS5063: The left-hand side of a '&&' expression cannot have type 'void'.
+tests/cases/conformance/expressions/binaryOperators/logicalAndOperator/logicalAndOperatorWithEveryType.ts(96,11): error TS5063: The left-hand side of a '&&' expression cannot have type 'void'.
+tests/cases/conformance/expressions/binaryOperators/logicalAndOperator/logicalAndOperatorWithEveryType.ts(107,11): error TS5063: The left-hand side of a '&&' expression cannot have type 'void'.
+tests/cases/conformance/expressions/binaryOperators/logicalAndOperator/logicalAndOperatorWithEveryType.ts(118,11): error TS5063: The left-hand side of a '&&' expression cannot have type 'void'.
+
+
+==== tests/cases/conformance/expressions/binaryOperators/logicalAndOperator/logicalAndOperatorWithEveryType.ts (10 errors) ====
+    // The && operator permits the operands to be of any type and produces a result of the same
+    // type as the second operand.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/logicalAndOperatorWithEveryType.errors.txt]: This might need to be updated. Or should I split this test up?

Comment (sandersn) [inline tests/baselines/reference/logicalAndOperatorWithEveryType.errors.txt]: yes

**Comment by sandersn** [comment]
Why didn't we check this before? This seems fundamental to the type of binary logical operators.

**Comment by jeffreymorlan** [comment]
`void` expressions should also be disallowed on the left side of `? :` or the right side of `!`, I think.

**Comment by DanielRosenwasser** [comment]
@jeffreymorlan that's not a bad idea either. Thanks!

**Comment by RyanCavanaugh** [comment]
:+1: and :white_check_mark:

**Comment by mhegazy** [comment]
@DanielRosenwasser can you update the PR.

**Comment by mhegazy** [comment]
@DanielRosenwasser can you update and merge this time :)

**Comment by mhegazy** [comment]
Thanks for your contribution. This PR has failing CI tests and can not be merged in at the time being. For housekeeping purposes we are closing stale PRs. If you'd still like to merge this code in, please open a new PR that has been merged and rebased with the `master` branch.

**Pull request closed without merging**